be/src/storage/vectorized/compaction.cpp

```diff
@@ -182,9 +283,87 @@ Status Compaction::merge_rowsets(int64_t mem_limit, Statistics* stats_output) {
 
     auto char_field_indexes = ChunkHelper::get_char_field_indexes(schema);
 
-    while (true) {
-        bool bg_worker_stopped = ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped();
+    Status status;
+    bool bg_worker_stopped = ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped();
+    while (!bg_worker_stopped) {
+#ifndef BE_TEST
+        status = tls_thread_status.mem_tracker()->check_mem_limit("Compaction");
+        if (!status.ok()) {
+            LOG(WARNING) << "fail to execute compaction: " << status.message() << std::endl;
+            return status;
+        }
+#endif
+
+        chunk->reset();
+        status = reader.get_next(chunk.get());
+        if (!status.ok()) {
+            if (status.is_end_of_file()) {
+                break;
+            } else {
+                LOG(WARNING) << "reader get next error. tablet=" << _tablet->full_name()
+                             << ", err=" << status.to_string();
+                return Status::InternalError(fmt::format("reader get_next error: {}", status.to_string()));
+            }
+        }
+
+        ChunkHelper::padding_char_columns(char_field_indexes, schema, _tablet->tablet_schema(), chunk.get());
+
+        OLAPStatus olap_status = _output_rs_writer->add_chunk(*chunk);
+        if (olap_status != OLAP_SUCCESS) {
+            LOG(WARNING) << "writer add_chunk error, err=" << olap_status;
+            return Status::InternalError("writer add_chunk error.");
+        }
+        output_rows += chunk->num_rows();
+
+        bg_worker_stopped = ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped();
+    }
+
+    if (bg_worker_stopped) {
+        return Status::InternalError("Process is going to quit. The compaction should be stopped as soon as possible.");
+    }
+
+    if (stats_output != nullptr) {
+        stats_output->output_rows = output_rows;
+        stats_output->merged_rows = reader.merged_rows();
+        stats_output->filtered_rows = reader.stats().rows_del_filtered;
+    }
+
+    OLAPStatus olap_status = _output_rs_writer->flush();
+    if (olap_status != OLAP_SUCCESS) {
+        LOG(WARNING) << "failed to flush rowset when merging rowsets of tablet " + _tablet->full_name()
+                     << ", err=" << olap_status;
+        return Status::InternalError("failed to flush rowset when merging rowsets of tablet error.");
+    }
+
+    return Status::OK();
+}
+
+Status Compaction::merge_rowsets_vertically(Statistics* stats_output) {
+    TRACE_COUNTER_SCOPE_LATENCY_US("merge_rowsets_latency_us");
+    std::unique_ptr<RowSourceMaskBuffer> mask_buffer =
+            std::make_unique<RowSourceMaskBuffer>(_tablet->tablet_id(), _tablet->data_dir()->path());
+    std::unique_ptr<std::vector<RowSourceMask>> source_masks = std::make_unique<std::vector<RowSourceMask>>();
+    for (size_t i = 0; i < _column_groups.size(); ++i) {
+        bool is_key = (i == 0);
+        if (!is_key) {
+            mask_buffer->flip();
```

Comment:
If there are multiple non-key column groups, `mask_buffer->flip()` will be called multiple times, does it ok?

Reply:
Need to read mask buffer from the beginning when starting a new group.